proposals/0032-middleware-api.md

```diff
@@ -0,0 +1,469 @@
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 
+- Reference Issues: 
+  - https://github.com/withastro/roadmap/issues/531
+- Implementation PR: https://github.com/withastro/astro/pull/6721
+
+# Summary
+
+Introduce a middleware to Astro, where you can define code that runs on every request. 
+This API should work regardless of the rendering mode (SSG or SSR) or adapter used. 
+Also introduces a simple way to share request-specific data between proposed middleware, 
+API routes, and `.astro` routes.
+
+# Example
+
+The middleware pattern is useful to read `Request` and manipulate the `Response`, other than
+set and share information across endpoints and pages.
+
+
+For example, here's an example of how a user can share some information across routes:
+```js
+export const onRequest = (context, next) => {
+    if (!context.request.url.endsWith("/")) {
+        context.locals.isIndex = false;
+    } else {
+        context.locals.isIndex = true;
+    }
+}
+```
+
+Or, set some logic to make redirects:
+```js
+const redirects = new Map([
+    ["/old-1", "/new-1"],
+    ["/old-1", "/new-1"]
+])
+
+export const onRequest = (context, next) => {
+    for (const [oldRoute, newRoute] of redirects) {
+        if (context.request.url.endsWith(oldRoute)) {
+            return context.redirect(newRoute, 302);
+        }
+    }
+}
+```
+
+# Background & Motivation
+
+Middleware has been one of the most heavily requested feature in Astro. 
+It's useful for handling common tasks like auth guards and setting cache headers. 
+
+For me, it would make handling authentication much easier.
+
+
+# Goals
+
+- Provide a way intercept requests and responses, allowing users to set cookies and headers
+- Works both in SSR and SSG mode
+- Allow users to use community-created middlewares (libraries)
+  - Make available via integrations API.
+- Provide an API for request-specific data
+- Non-Node runtimes specific APIs. ie. Cloudflare Durable Objects.
+  - Add middleware from adapter.
+
+# Non-Goals
+
+- Route specific middleware, middlewares that are run **only on specific** routes
+
+# Detailed Design
+
+## Changes from the Stage 2 proposal
+
+- `resolve` has been renamed to `next`;
+- `next` doesn't accept an `APIContext` to work;
+- `locals` values need to be serializable to avoid the introduction of
+  non-user code from third-party libraries that can run scripts;
+- `middleware` export function has been renamed `onRequest`. This name change
+  has two benefits:
+  1. It shows intent and explains when this function is called;
+  2. It allows adding more functions with the `on*` prefix in the future, which could show the intent
+     of when the function is called in the Astro route life cycle;
+
+## Implementation instructions
+
+To define a middleware, a user must create a physical file under the `src/` folder, called `middleware.js`.
```

Comment:
To be precise, this should be `config.srcDir`, see https://docs.astro.build/en/reference/configuration-reference/#srcdir